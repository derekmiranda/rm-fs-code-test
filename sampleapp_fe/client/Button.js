import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ fetchData, text }) => (
  <button onClick={fetchData}>{text}</button>
)

Button.propTypes = {
  fetchData: PropTypes.func,
  text: PropTypes.string,
}

export default Button;
