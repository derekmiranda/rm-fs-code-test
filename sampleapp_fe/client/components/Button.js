import React from 'react';
import PropTypes from 'prop-types';

const style = {
  fontSize: '1em',
  paddingBottom: '0.5em',
}

const Button = (props) => {
  const { fetchData, text } = props;
  return (
    <button onClick={fetchData} style={style}>{text}</button>
  )
}

Button.propTypes = {
  fetchData: PropTypes.func,
  text: PropTypes.string,
}

export default Button;
