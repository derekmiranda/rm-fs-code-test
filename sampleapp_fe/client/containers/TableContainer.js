import React from 'react';
import PropTypes from 'prop-types';
import orderedFieldInfo from '../utils/orderedFieldInfo';
import Table from '../components/Table';

const getPropertyValuesInOrder = (property) => {
  const propertyValues = orderedFieldInfo.map(
    fieldInfo => property[fieldInfo.rawName]
  )
  return propertyValues;
}

const TableContainer = ({ properties }) => {
  const labels = orderedFieldInfo.map(fieldInfo => fieldInfo.label);
  const rowValues = properties.map(property => getPropertyValuesInOrder(property));

  return <Table rowValues={rowValues} labels={labels}/>
}

TableContainer.propTypes = {
  properties: PropTypes.array,
}

export default TableContainer;
