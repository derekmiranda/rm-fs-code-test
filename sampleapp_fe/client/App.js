import React, { Component } from 'react';
import Button from './Button';

const DATA_URL = 'http://localhost:9000/data';

class App extends Component {
  constructor() {
    super();
    this.state = {
      properties: [],
    }
    this._fetchData = async () => {
      const res = await fetch(DATA_URL);
      const properties = await res.json();
      this.setState({
        properties,
      })
    }
  }

  render() {
    return (
      <main>
        <h1>Hello World!</h1>
        <Button fetchData={this._fetchData} text='Fetch Properties' />
        <ul>{this.state.properties && 
          this.state.properties.map((property, i) => <li key={i}>{JSON.stringify(property)}</li>)
        }</ul>
      </main>
    )
  }
}

export default App;
