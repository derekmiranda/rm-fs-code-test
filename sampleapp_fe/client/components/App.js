import React, { Component } from 'react';
import Button from './Button';
import TableContainer from '../containers/TableContainer';
import processProperty from '../utils/processProperty';

const DATA_URL = 'http://localhost:9000/data';

const style = {
  fontFamily: 'sans-serif',
}

class App extends Component {
  constructor() {
    super();
    this.state = {
      properties: [],
    }
    this._fetchData = async () => {
      const res = await fetch(DATA_URL);
      const rawProperties = await res.json();
      const properties = rawProperties.map(processProperty);
      this.setState({
        properties,
      })
    }
  }

  render() {
    const table = this.state.properties.length ?
      <TableContainer properties={this.state.properties} /> :
      null;

    return (
      <main style={style}>
        <h1>California Properties</h1>
        <Button fetchData={this._fetchData} text='Fetch Properties' />
        {table}
      </main>
    )
  }
}

export default App;
