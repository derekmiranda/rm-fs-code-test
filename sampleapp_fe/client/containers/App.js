import React, { Component } from 'react';
import Button from '../components/Button';
import TableContainer from './TableContainer';
import processProperty from '../utils/processProperty';

const DATA_URL = 'http://localhost:9000/data';

const style = {
  fontFamily: 'sans-serif',
}

class App extends Component {
  constructor() {
    super();
    this.state = {
      properties: [],
      fetchState: '',
    }
    const setStatePromise = (stateChange) => {
      const component = this;
      return new Promise((resolve, reject) => {
        component.setState(stateChange, resolve);
      });
    }

    this._fetchData = async () => {
      await setStatePromise({ fetchState: 'Fetching data...' });
      const res = await fetch(DATA_URL);
      await setStatePromise({ fetchState: '' });
      const rawProperties = await res.json();
      const properties = rawProperties.map(processProperty);
      this.setState({
        properties,
      })
    }
  }

  render() {
    const table = this.state.properties.length ?
      <TableContainer properties={this.state.properties} /> :
      null;
    const fetchState = this.state.fetchState && (
      <p>{this.state.fetchState}</p>
    );

    return (
      <main style={style}>
        <h1>California Properties</h1>
        <Button fetchData={this._fetchData} text='Fetch Properties' />
        {fetchState}
        {table}
      </main>
    )
  }
}

export default App;
