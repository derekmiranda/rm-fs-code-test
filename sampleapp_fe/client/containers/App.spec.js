import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import TableContainer from './TableContainer';
import Button from './Button';

describe('<App />', () => {
  it('should render necessary components', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.containsAllMatchingElements([
      <Button />,
      <TableContainer />
    ])).toBe(true)
  })
})