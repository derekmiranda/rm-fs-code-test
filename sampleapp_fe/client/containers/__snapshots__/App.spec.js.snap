// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
<main
  style={
    Object {
      "fontFamily": "sans-serif",
    }
  }
>
  <h1>
    California Properties
  </h1>
  <div
    style={
      Object {
        "paddingBottom": "0.5em",
      }
    }
  >
    <button
      onClick={[Function]}
      style={
        Object {
          "fontSize": "1em",
          "paddingBottom": "0.5em",
        }
      }
    >
      Fetch Properties
    </button>
  </div>
  
</main>
`;
